A certain combination of iterators and procs crashes the compiler

### Code to reproduce
```nim
import std / sequtils

iterator combinations*[T](s: seq[T], r: Positive): seq[T] =
  yield @[s[0], s[1]]

iterator pairwise*[T](s: openArray[T]): seq[T] =
  yield @[s[0], s[1]]

proc checkSpecialSubset5(s: seq[int]): bool =
  toSeq(
    toSeq(
      s.combinations(2)
    ).map(proc(a: auto): int = a[0]).pairwise()
  ).any(proc(a: auto): bool = true)

echo checkSpecialSubset5 @[1, 2]
```

### Current Output
```
> nim c -r "d:\[REDACTED].nim"
Hint: used config file 'C:\Users\EyeCon\.choosenim\toolchains\nim-#devel\config\nim.cfg' [Conf]
Hint: used config file 'C:\Users\EyeCon\.choosenim\toolchains\nim-#devel\config\config.nims' [Conf]
Hint: used config file 'd:\prog\Nim\pe\config.nims' [Conf]
...............................................................
[REDACTED](1, 12) Warning: imported and not used: 'algorithm' [UnusedImport]
CC: pe0103.nim
C:\Users\EyeCon\nimcache\pe0103_d\@mpe0103.nim.c: In function 'checkSpecialSubset5__pe48494851_13':
C:\Users\EyeCon\nimcache\pe0103_d\@mpe0103.nim.c:629:55: error: redeclaration of 'resultX60gensym6_' with no linkage
  629 |                 tySequence__PSP8snSsRoFs9cDiya9bd7UQ* resultX60gensym6_;
      |                                                       ^~~~~~~~~~~~~~~~~
C:\Users\EyeCon\nimcache\pe0103_d\@mpe0103.nim.c:626:55: note: previous declaration of 'resultX60gensym6_' with type 'tySequence__PSP8snSsRoFs9cDiya9bd7UQ *'      
  626 |                 tySequence__PSP8snSsRoFs9cDiya9bd7UQ* resultX60gensym6_;
      |                                                       ^~~~~~~~~~~~~~~~~
C:\Users\EyeCon\nimcache\pe0103_d\@mpe0103.nim.c:633:55: error: redeclaration of 'resultX60gensym6_' with no linkage
  633 |                 tySequence__PSP8snSsRoFs9cDiya9bd7UQ* resultX60gensym6_;
      |                                                       ^~~~~~~~~~~~~~~~~
C:\Users\EyeCon\nimcache\pe0103_d\@mpe0103.nim.c:629:55: note: previous declaration of 'resultX60gensym6_' with type 'tySequence__PSP8snSsRoFs9cDiya9bd7UQ *'      
  629 |                 tySequence__PSP8snSsRoFs9cDiya9bd7UQ* resultX60gensym6_;
      |                                                       ^~~~~~~~~~~~~~~~~
C:\Users\EyeCon\nimcache\pe0103_d\@mpe0103.nim.c:636:55: error: redeclaration of 'resultX60gensym6_' with no linkage
  636 |                 tySequence__PSP8snSsRoFs9cDiya9bd7UQ* resultX60gensym6_;
      |                                                       ^~~~~~~~~~~~~~~~~
C:\Users\EyeCon\nimcache\pe0103_d\@mpe0103.nim.c:633:55: note: previous declaration of 'resultX60gensym6_' with type 'tySequence__PSP8snSsRoFs9cDiya9bd7UQ *'      
  633 |                 tySequence__PSP8snSsRoFs9cDiya9bd7UQ* resultX60gensym6_;
      |                                                       ^~~~~~~~~~~~~~~~~
Error: execution of an external compiler program 'gcc.exe -c  -w -fmax-errors=3 -mno-ms-bitfields -O3 -fno-strict-aliasing -fno-ident   -IC:\Users\EyeCon\.choosenim\toolchains\nim-#devel\lib -Id:\prog\Nim\pe -o C:\Users\EyeCon\nimcache\pe0103_d\@mpe0103.nim.c.o C:\Users\EyeCon\nimcache\pe0103_d\@mpe0103.nim.c' failed with exit code: 1
```

### Expected Output
That the file is compiled without problems

### Possible Solution
From the discussion in Discord:
> Solitude: i can change inner iterator to seq and it will still break
> Solitude: but pairwise needs to be openArray to not work

### Additional Information

```
>nim -v
Nim Compiler Version 1.7.1 [Windows: amd64]
Compiled at 2022-03-27
Copyright (c) 2006-2022 by Andreas Rumpf

active boot switches: -d:release

# make sure to include the git hash if not using a tagged release
```

For easier testing, here a slightly different version with templates:
```nim
import std / [algorithm, sequtils, sugar]

template bar(): untyped =
    toSeq(toSeq(s.combinations(2)).map(a => a).pairwise()).any(a => true) # Exhibits the bug
    # toSeq(toSeq(s.combinations(2)).map(proc(p: auto): auto = p).pairwise()).any(a => true) # Exhibits the bug
    # toSeq(toSeq(s.combinations(2)).pairwise()).any(a => true) # Everything OK

iterator combinations*[T](s: openArray[T], r: Positive): seq[T] =
  yield @[s[0], s[1]]
  yield @[s[0], s[1]]

iterator pairwise*[T](s: openArray[T]): seq[T] =
  yield @[s[0], s[1]]
  yield @[s[0], s[1]]

proc checkSpecialSubset5(s: seq[int]): bool = bar()

when isMainModule:
    let s = @[3,4,5,6]
    echo bar() # Works
    if bar(): # Works
        echo "Foo"
    echo checkSpecialSubset5(s) # With the bug: doesn't work
```

--- a/compiler/transf.nim
+++ b/compiler/transf.nim
@@ -635,8 +635,14 @@ proc putArgInto(arg: PNode, formal: PType): TPutArgInto =
   if formal.kind == tyTypeDesc: return paDirectMapping
   if skipTypes(formal, abstractInst).kind in {tyOpenArray, tyVarargs}:
     case arg.kind
-    of nkStmtListExpr, nkCall:
+    of nkStmtListExpr:
       return paComplexOpenarray
+    of nkCall:
+      if skipTypes(arg.typ, abstractInst).kind in {tyOpenArray, tyVarargs}:
+        # XXX incorrect, causes #13417 when `arg` has side effects.
+        return paDirectMapping
+      else:
+        return paComplexOpenarray
     of nkBracket:
       return paFastAsgnTakeTypeFromArg
     else:

--- a/compiler/transf.nim
+++ b/compiler/transf.nim
@@ -635,7 +635,7 @@ proc putArgInto(arg: PNode, formal: PType): TPutArgInto =
   if formal.kind == tyTypeDesc: return paDirectMapping
   if skipTypes(formal, abstractInst).kind in {tyOpenArray, tyVarargs}:
     case arg.kind
-    of nkStmtListExpr:
+    of nkStmtListExpr, nkCall:
       return paComplexOpenarray
     of nkBracket:
       return paFastAsgnTakeTypeFromArg
@@ -803,7 +803,7 @@ proc transformFor(c: PTransf, n: PNode): PNode =
       stmtList.add(newAsgnStmt(c, nkFastAsgn, temp, addrExp, true))
       newC.mapping[formal.itemId] = newDeref(temp)
     of paComplexOpenarray:
-      # arrays will deep copy here (pretty bad).
+      # XXX arrays will deep copy here (pretty bad).
       var temp = newTemp(c, arg.typ, formal.info)
       addVar(v, temp)
       stmtList.add(newAsgnStmt(c, nkFastAsgn, temp, arg, true))

--- a/tests/iter/titeropenarray.nim
+++ b/tests/iter/titeropenarray.nim
@@ -1,0 +1,32 @@
+block: # issue #13417
+  var s: seq[int] = @[]
+  proc p1(): seq[int] =
+    s.add(3)
+    @[1,2]
+
+  iterator ip1(v: openArray[int]): auto =
+    for x in v:
+      yield x
+
+  for x in ip1(p1()):
+    s.add(x)
+
+  doAssert s == @[3, 1, 2]
+
+import std / sequtils
+
+block: # issue #19703
+  iterator combinations[T](s: seq[T], r: Positive): seq[T] =
+    yield @[s[0], s[1]]
+
+  iterator pairwise[T](s: openArray[T]): seq[T] =
+    yield @[s[0], s[0]]
+
+  proc checkSpecialSubset5(s: seq[int]): bool =
+    toSeq(
+      toSeq(
+        s.combinations(2)
+      ).map(proc(a: auto): int = a[0]).pairwise()
+    ).any(proc(a: auto): bool = a == @[s[0], s[0]])
+
+  doAssert checkSpecialSubset5 @[1, 2]

